src/libraries/System.Private.CoreLib/src/System/Text/UnicodeEncoding.cs

```diff
@@ -1535,7 +1535,7 @@ internal sealed override unsafe int GetCharCount(byte* bytes, int count, Decoder
                     }
 
                     // Valid surrogate pair, add our lastChar (will need 2 chars)
-                    if (chars >= charEnd - 1)
+                    if (chars >= charEnd - 1 || chars == charEnd)
```

Comment:
I am uncertain about the rationale behind this modification. when `chars == charEnd`, this implies that the condition `chars >= charEnd - 1` remains true. The added part is not changing anything. am I missing something?

Reply:
`charEnd` can be null. `charEnd - 1` will underflow in that case, and the first condition won't be true.

Reply:
If this is the case, should we have a simple check in the top of the method to check that and just throw if we have any bytes to encode?

Reply:
I am seeing `charEnd - chars` is used in some other place in the method too.

Reply:
> If this is the case, should we have a simple check in the top of the method to check that and just throw if we have any bytes to decode?

It would not be correct, and it would be a breaking change. Just because we have some input bytes does not mean that we end up outputting any chars.

Reply:
But I am seeing we have the assert 

            Debug.Assert(chars is not null, "[UnicodeEncoding.GetChars]chars!=null");


on the top of the method. Isn't doing the exact same check? I know this is done in debug build but I would expect the behavior should be consistent.

Reply:
The higher-level encoding methods convert `null` pointers to `(char*)1` for backward compatibility (look for `MemoryMarshal.GetNonNullPinnableReference` use under System.Text).

`(char*)1` is not null. It explains why this assert does not fire.

Values of the pointers in the broken case that this change is fixing are:
```
char* chars = (char*)1;
char* charEnd = chars + 0; // This is (char*)1

... charEnd - 1 ... // This underflows to 0xFFFFFFF... . 
```

Reply:
I think it is worth adding a comment with your example. Otherwise anyone reading this code will not understand what is going on. Thanks @jkotas for explanation.

Reply:
What would you like to see the comment to say?

I think the new code is easy to understand, without additional comments:
```
// Valid surrogate pair, add our lastChar (will need 2 chars)
if (charEnd - chars < 2)
```

When one has start pointer, end pointer and required length, the right pattern to use to check for sufficient space is `end_pointer - start_pointer >= required_length`. Other variants like `end_pointer - required_length >= start_pointer` or `end_pointer >= start_pointer + required_length` are prone to integer overflows. It is one of those things that is mentioned in SDL training.

Reply:
If it is not worth the comment, that is ok. It was not obvious to me we can get char* is null or `1` value as I am not expecting this to happen in this code base. The comment in my mind is something like `Exercise caution when manipulating pointers to prevent potential underflow issues. In certain situations, the char* can be equal to 1 and may also have a length of zero.`

I'll leave it to you to decide. I am ok either way.

Reply:
@manandre Do you have a preference?